extensions/amp-form/0.1/form-validators.js

```diff
@@ -0,0 +1,347 @@
+        break;
+      }
+    }
+  }
+
+  /** @override */
+  onBlur(unusedEvent) {
+    validationBubble.hide();
+  }
+
+  /** @override */
+  onInput(event) {
+    if (event.target.checkValidity()) {
+      event.target.removeAttribute('aria-invalid');
+      validationBubble.hide();
+    } else {
+      event.target.setAttribute('aria-invalid', 'true');
+      validationBubble.show(event.target, event.target.validationMessage);
+    }
+  }
+}
+
+
+class AbstractCustomValidator extends FormValidator {
+
+  /**
+   * @param {!Element} input
+   */
+  reportInput(input) {
+    const invalidType = getInvalidType(input);
+    if (invalidType) {
+      this.showValidationFor(input, invalidType);
+    }
+  }
+
+  /**
+   * Hides all validation messages.
+   */
+  hideAllValidations() {
+    const inputs = this.form.querySelectorAll('input,select,textarea');
+    for (let i = 0; i < inputs.length; i++) {
+      this.hideValidationFor(inputs[i]);
+    }
+  }
+
+  /**
+   * @param {!Element} input
+   * @param {!string} invalidType
+   * @return {?Element}
+   */
+  getValidationFor(input, invalidType) {
+    if (!input.id) {
+      return null;
+    }
+
+    return this.doc.querySelector(
+        `[visible-when-invalid=${invalidType}][validation-for=${input.id}]`);
+  }
+
+  /**
+   * @param {!Element} input
+   * @param {string} invalidType
+   */
+  showValidationFor(input, invalidType) {
+    const validation = this.getValidationFor(input, invalidType);
+    if (!validation) {
+      return;
+    }
+
+    if (!validation.textContent.trim()) {
+      validation.textContent = input.validationMessage;
+    }
+
+    input.setAttribute('aria-invalid', 'true');
+    validation.classList.add('visible');
+  }
+
+  /**
+   * @param {!Element} input
+   * @return {boolean} Whether a validation was visible and was successfully hidden.
+   */
+  hideValidationFor(input) {
+    const visibleValidation = this.getVisibleValidationFor(input);
+    if (!visibleValidation) {
+      return false;
+    }
+
+    input.removeAttribute('aria-invalid');
+    visibleValidation.classList.remove('visible');
+    return true;
+  }
+
+  /**
+   * @param {!Element} input
+   * @return {?Element}
+   */
+  getVisibleValidationFor(input) {
+    if (!input.id) {
+      return null;
+    }
+    return this.doc.querySelector(
+        `[visible-when-invalid][validation-for=${input.id}].visible`);;
+  }
+
+  /**
+   * Whether an input should validate after an interaction.
+   * @param {!Element} input
+   * @return {boolean}
+   */
+  shouldValidateOnInteraction(input) {
+    return !!this.getVisibleValidationFor(input);
+  }
+
+  /**
+   * @param {!Event} event
+   */
+  onInteraction(event) {
+    const input = /** @type {!Element} */ (event.target);
+    const shouldValidate = this.shouldValidateOnInteraction(input);
+    this.hideValidationFor(input);
+    if (shouldValidate && !event.target.checkValidity()) {
```

Comment:
We can just `reportInput` without checking validity but my guess `checkValidity` is cheap (will talk to Chrome team to check) that way we can just rely on the correctness of `input.validity` state.